### ISSUE TITLE
implicit None return checks OK against type hint of bool or Iterator[int]

### ISSUE THREAD

**Issue by dckc** [python/mypy#8054]
> Are you reporting a bug, or opening a feature request?

bug

(or at least: where is this documented?)

> Please insert below the code you are checking with mypy

```python
import typing as ty

def each():
    # type: () -> ty.Iterator[int]
    pass


for x in each():
    print(x)
```

or even just:

```python
def gt(a, b):
    # type: (int, int) -> bool
    "blather..."
```

> What is the actual behavior/output?
```
Success: no issues found in 1 source file
```

> What is the behavior/output you expect?

An error; either "None found where Iterator[int] expected" or "Not all paths return a value" or some such.

Note that python doesn't like this code:

```
python ty1.py
Traceback (most recent call last):
  File "ty1.py", line 8, in <module>
    for x in each():
TypeError: 'NoneType' object is not iterable
```

> What are the versions of mypy and Python you are using?

mypy 0.730
Python 3.7.3

>  Do you see the same issue after installing mypy from Git master?

yes. (mypy-0.760+dev.f9955ae1101cd7d1b38c39e8fba77f1d74986a02 , Python 3.8.0)

> What are the mypy flags you are using? (For example --strict-optional)

I was originally using:

```
MYPYPATH=/home/dconnolly/heron/.direnv/python-2.7.15+/lib/python2.7/site-packages
~/opt/miniconda3/envs/py3tr/bin/mypy -2 \
	--follow-imports=skip \
	--disallow-untyped-calls --disallow-untyped-defs \
		ty1.py
```

but when I tested mypy from git, I used (a) no flags and (b) `--strict`.

**Comment by JukkaL** [comment]
Yeah, this kind of unfortunate. The motivation for this is that it's common to declare abstract methods with only `pass` in the body, and even if the method is declared to return a non-`None` value. However, we probably should have at least a strictness setting to catch this issue (if we can't make it the default). Once we support enabling/disabling individual error codes, we could perhaps use this mechanism to implement the option. Surprisingly, I couldn't find an existing issue about this.

**Comment by ilevkivskyi** [comment]
Duplicate of https://github.com/python/mypy/issues/2350

**Comment by dckc** [comment]
Ah... I do get "Missing return statement" for non-trivial function bodies such as:

```python
def gt(a, b):
    #type: (int, int) -> bool
    x = a + b
```

So this is a much more narrow issue than I thought.
